README.md
# ⬇️ node-rutorrent-promise

[![version](https://badgen.net/npm/v/rutorrent-promise)](https://www.npmjs.com/package/rutorrent-promise)
[![dependencies](https://badgen.net/david/dep/ctessier/node-rutorrent-promise)](https://www.npmjs.com/package/rutorrent-promise)
[![size](https://badgen.net/bundlephobia/min/rutorrent-promise)](https://bundlephobia.com/result?p=rutorrent-promise)
[![downloads](https://badgen.net/npm/dt/rutorrent-promise)](https://www.npmjs.com/package/rutorrent-promise)
![license](https://badgen.net/npm/license/rutorrent-promise)

__Interact with ruTorrent via promises__ 👌

This was inspired by [Grant](https://github.com/grantholle)'s [transmission](https://www.npmjs.com/package/transmission-promise) library. It provides methods to communicate with ruTorrent client using promises. It works thanks to the [HTTPRPC](https://github.com/Novik/ruTorrent/wiki/PluginHTTPRPC) plugin.

## Installation

```bash
npm i rutorrent-promise --save
```

```javascript
const RuTorrent = require('rutorrent-promise');

const rutorrent = new RuTorrent({
  host: 'localhost', // default: localhost
  port: 80,          // default: 80
  path: '',          // default: /rutorrent
  ssl: true,         // default: false
  username: '',      // default: none
  password: '',      // default: none
});
```

## Available fields

An array of fields can be passed to the `get` and `addFile` methods in order to retrieve specific values from the torrents. The list is taken from the available properties provided by the [HTTPRPC ruTorrent plugin](https://github.com/Novik/ruTorrent/blob/master/plugins/httprpc/action.php#L90). Here it is:

- `d.is_open`
- `d.is_hash_checking`
- `d.is_hash_checked`
- `d.get_state`
- `d.get_name`
- `d.get_size_bytes`
- `d.get_completed_chunks`
- `d.get_size_chunks`
- `d.get_bytes_done`
- `d.get_up_total`
- `d.get_ratio`
- `d.get_up_rate`
- `d.get_down_rate`
- `d.get_chunk_size`
- `d.get_custom1`
- `d.get_peers_accounted`
- `d.get_peers_not_connected`
- `d.get_peers_connected`
- `d.get_peers_complete`
- `d.get_left_bytes`
- `d.get_priority`
- `d.get_state_changed`
- `d.get_skip_total`
- `d.get_hashing`
- `d.get_chunks_hashed`
- `d.get_base_path`
- `d.get_creation_date`
- `d.get_tracker_focus`
- `d.is_active`
- `d.get_message`
- `d.get_custom2`
- `d.get_free_diskspace`
- `d.is_private`
- `d.is_multi_file`

## Methods

### `get(fields = [])`

List all the torrents.

```javascript
rutorrent
  .get(['d.get_name', 'd.get_custom1', 'd.get_size_bytes'])
    .then((data) => {
      console.log(data);
      // [
      //   {
      //     hashString: '286D2E5B4F8369855328336AC1263AE02A7A60D5',
      //     'd.get_name': 'ubuntu-18.04.4-desktop-amd64.iso',
      //     'd.get_custom1': 'linux-distro',
      //     'd.get_size_bytes': '2126544896'
      //   },
      //   {
      //     hashString: 'EE55335F2ACDE309FA645FAB11C04750D7E45FA1',
      //     'd.get_name': 'ubuntu-16.04.6-desktop-amd64.iso',
      //     'd.get_custom1': 'linux-distro',
      //     'd.get_size_bytes': '1664614400'
      //   }
      // ]
    })
    .catch((err) => {
      // something happened...
    });
```

### `addFile(file, options = {})`

Add a new torrent from a file.

```javascript
rutorrent
  .addFile(fs.readFileSync('/path/to/ubuntu-18.04.4-desktop-amd64.iso.torrent'), {
    label: 'linux-distro',
    destination: '/downloads',
  })
    .then(() => {
      // torrent was added!
    })
    .catch((err) => {
      // something happenned...
    });
```

### `addUrl(url, options = {})`

Add a new torrent from an url.

```javascript
rutorrent
  .addUrl('http://releases.ubuntu.com/18.04/ubuntu-18.04.4-desktop-amd64.iso.torrent', {
    label: 'linux-distro',
    destination: '/downloads',
  })
    .then(() => {
      // torrent was added!
    })
    .catch((err) => {
      // something happenned...
    });
```

### `delete(hash, deleteTiedFiles = true)`

Delete a torrent.

```javascript
rutorrent
  .delete('286D2E5B4F8369855328336AC1263AE02A7A60D5')
    .then(() => {
      // torent has been deleted!
    })
    .catch((err) => {
      // something happened...
    });
```

## Contributing

👨‍💻👩‍💻

Feel free to contribute and help me add more methods to interact with ruTorrent.

To run the unit tests, create an `.env` file in the `test` directory and fill it in with your client host and credentials:

```bash
$ cp test/.env.sample test/.env
$ cat test/.env
HOST=YOUR_CLIENT_HOST
PORT=YOUR_CLIENT_PORT
PREFIX_PATH=YOUR_CLIENT_PREFIX_PATH # (default is /rutorrent)
SSL=true                            # (https or http)
LOGIN=YOUR_CLIENT_LOGIN
PASSWORD=YOUR_CLIENT_PASSWORD

$ npm i && npm t
```


CHANGELOG.md
# Release Notes

## [v2.0.0](https://github.com/ctessier/node-rutorrent-promise/compare/v1.3.0...v2.0.0)

> Released 2020/03/29

:warning: This release introduces breaking changes from 1.3.0.

### Breaking changes

The methods to add a torrent no longer return details on the added torrent. This feature was time consuming as it required two requests. It was also not working as expected when adding a torrent that already existed. Check the [documentation](https://github.com/ctessier/node-rutorrent-promise/blob/v2.0.0/README.md) for more details.

If you still wish to get details on a torrent file or magnet, have a look at the [webtorrent/parse-torrent](https://github.com/webtorrent/parse-torrent) library.

### Changed

- Improvement was made on the library weight. [axios](https://github.com/axios/axios) has been replaced by [node-fetch](https://github.com/node-fetch/node-fetch) which drastically reduces the size of the package.


## [v1.3.0](https://github.com/ctessier/node-rutorrent-promise/compare/v1.2.0...v1.3.0)

> Released 2020/03/20

### Added

- Possibility to add a torrent via an URL ([#13](https://github.com/ctessier/node-rutorrent-promise/pull/13))

### Fixed

- Conflict with `USERNAME` environment variable on Linux ([#14](https://github.com/ctessier/node-rutorrent-promise/pull/14))

## [v1.2.0](https://github.com/ctessier/node-rutorrent-promise/compare/v1.1.1...v1.2.0)

> Released 2020/03/15

### Added

- Possibility to delete a torrent ([#8](https://github.com/ctessier/node-rutorrent-promise/pull/8))
- Upgrade Axios version to 0.19.2 ([#10](https://github.com/ctessier/node-rutorrent-promise/pull/10))

### Fixed

- Conflict with PATH environment variable ([#9](https://github.com/ctessier/node-rutorrent-promise/pull/9))

## [v1.1.1](https://github.com/ctessier/node-rutorrent-promise/compare/v1.1.0...v1.1.1)

> Released 2020/02/19

### Fixed

- fix option property for providing a custom path (Thanks [@ShaunLWM](https://github.com/ShaunLWM) [#5](https://github.com/ctessier/node-rutorrent-promise/pull/5))


## [v1.1.0](https://github.com/ctessier/node-rutorrent-promise/compare/v1.0.0...v1.1.0)

> Released 2019/11/02

### Breaking changes

- `totalBytes`, `bytesDone` and `label` are no longer returned from the `get` method. You must now specified a list of fields to retrieve (see documentation)

### Added

- possibility to specify wanted fields when retrieving torrents ([#1](https://github.com/ctessier/node-rutorrent-promise/pull/1))

## v1.0.0

> Released 2019/11/01


test/rutorrent.spec.js
const fs = require('fs');
const path = require('path');
const http = require('http');
const expect = require('chai').expect;

require('dotenv').config({
  path: path.resolve(`${process.cwd()}/test`, '.env'),
});

if (!fs.existsSync(path.resolve('tmp'))) {
  fs.mkdirSync(path.resolve('tmp'))
}

const RuTorrent = require('../');
const options = {};

const dataSet = [{
  name: 'ubuntu-18.04.4-desktop-amd64.iso',
  hash: '286D2E5B4F8369855328336AC1263AE02A7A60D5',
  url: 'http://releases.ubuntu.com/18.04/ubuntu-18.04.4-desktop-amd64.iso.torrent',
  label: 'linux-distro',
}, {
  name: 'ubuntu-16.04.6-desktop-amd64.iso',
  hash: 'EE55335F2ACDE309FA645FAB11C04750D7E45FA1',
  url: 'http://releases.ubuntu.com/16.04/ubuntu-16.04.6-desktop-amd64.iso.torrent',
  label: 'linux-distro',
}];

if (process.env.HOST) {
  options.host = process.env.HOST;
}
if (process.env.PORT) {
  options.port = process.env.PORT;
}
if (process.env.PREFIX_PATH) {
  options.path = process.env.PREFIX_PATH;
}
if (process.env.SSL) {
  options.ssl = process.env.SSL;
}
if (process.env.LOGIN) {
  options.username = process.env.LOGIN;
}
if (process.env.PASSWORD) {
  options.password = process.env.PASSWORD;
}

describe('rutorrent', () => {
  let rutorrent;

  it('can instantiate a new instance', (done) => {
    try {
      rutorrent = new RuTorrent(options);
      done();
    } catch(err) {
      done(err);
    }
  });

  describe('methods', () => {
    it('get - returns the list of torrents', (done) => {
      rutorrent.get(['d.get_name', 'd.get_size_bytes', 'd.get_bytes_done', 'd.get_custom1'])
        .then((result) => {
          expect(result).to.be.an('array');
          if (result.length) {
            result.forEach((res) => {
              expect(res).to.have.property('hashString');
              expect(res).to.have.property('d.get_name');
              expect(res).to.have.property('d.get_size_bytes');
              expect(res).to.have.property('d.get_bytes_done');
              expect(res).to.have.property('d.get_custom1');
            })
          }
          done();
        })
        .catch(done);
    });

    it('addFile - should add torrent from file', (done) => {
      const destination = path.resolve('tmp', path.basename(dataSet[0].url));
      const writeStream = fs.createWriteStream(destination);
      http.get(dataSet[0].url, (response) => {
        response.pipe(writeStream);
        response.on('error', done);
        response.on('end', () => {
          rutorrent.addFile(fs.readFileSync(destination), { label: dataSet[0].label })
            .then(() => rutorrent.get(['d.get_name', 'd.get_custom1'])
              .then((results) => {
                const torrent = results.find(r => r.hashString === dataSet[0].hash);
                expect(torrent).to.eql({
                  hashString: dataSet[0].hash,
                  'd.get_name': dataSet[0].name,
                  'd.get_custom1': dataSet[0].label,
                });
                done();
              }))
            .catch(done);
        });
      });
    });

    it('addUrl - should add torrent from url', (done) => {
      rutorrent.addUrl(dataSet[1].url, { label: dataSet[1].label })
        .then(() => rutorrent.get(['d.get_name', 'd.get_custom1'])
          .then((results) => {
            const torrent = results.find(r => r.hashString === dataSet[0].hash);
            expect(torrent).to.eql({
              hashString: dataSet[0].hash,
              'd.get_name': dataSet[0].name,
              'd.get_custom1': dataSet[0].label,
            });
            done();
          }))
        .catch(done);
    });

    it('delete - should delete torrent from hash', (done) => {
      rutorrent.delete(dataSet[0].hash, true)
        .then(() => rutorrent.get()
          .then((results) => {
            expect(results.every(r => r.hashString !== dataSet[0].hash)).to.be.true;

            return rutorrent.delete(dataSet[1].hash, true)
              .then(() => rutorrent.get()
                .then((results) => {
                  expect(results.every(r => r.hashString !== dataSet[1].hash)).to.be.true;
                  done();
                }))
          }))
        .catch(done);
    });
  });

});


index.js
'use strict';

const qs = require('querystring');
const fetch = require('node-fetch');
const FormData = require('form-data');

const utils = require('./utils');

class RuTorrent {

  /**
   * RuTorrent class constructor.
   *
   * Available options:
   *   - host
   *   - port
   *   - path
   *   - ssl (bool)
   *   - username
   *   - password
   *
   * @param {object} options
   */
  constructor(options = {}) {
    const protocol = options.ssl === true ? 'https' : 'http';
    const host = options.host || 'localhost';
    const port = options.port || 80;
    const path = options.path || '/rutorrent';

    this.baseUrl = `${protocol}://${host}:${port}${path}`;

    if (options.username) {
      this.authorizationHeader = `Basic ${Buffer.from(`${options.username}:${options.password || ''}`).toString('base64')}`;
    }
  }

  /**
   * Makes a call to the ruTorrent server.
   *
   * @return {Promise}
   */
  callServer(options) {
    let headers = {
      authorization: this.authorizationHeader,
    };

    if (options.data instanceof FormData) {
      headers = {
        ...options.data.getHeaders(),
        ...headers,
      };
    }

    return fetch(`${this.baseUrl}${options.path}`, {
      method: 'POST',
      redirect: 'manual',
      headers,
      body: options.data,
    }).then((res) => {
      if (res.ok) {
        return res.text();
      }
      else if (res.status === 302) {
        if (res.headers.get('location').indexOf('Success') > 0) {
          return res.text();
        }
        throw new Error('Error received from RuTorrent');
      }
      else {
        throw new Error(res.statusText);
      }
    }).then((data) => {
      try {
        return JSON.parse(data);
      } catch (e) {
        return data;
      }
    });
  }

  /**
   * Adds a new torrent from a given file or buffer.
   *
   * Available options:
   *   - label
   *   - destination
   *
   * @param  {string|Buffer} file
   * @param  {object}        options
   * @return {Promise}
   */
  addFile(file, options = {}) {
    const params = new FormData();
    params.append('torrent_file', file, 'torrent');

    if (options.label) {
      params.append('label', options.label);
    }
    if (options.destination) {
      params.append('dir_edit', options.destination);
    }

    return this.callServer({
      path: '/php/addtorrent.php',
      data: params,
    });
  }

  /**
   * Adds a new torrent from a given url.
   *
   * Available options:
   *   - label
   *   - destination
   *
   * @param  {string}  url
   * @param  {object}  options
   * @return {Promise}
   */
  addUrl(url, options = {}) {
    const params = new FormData();
    params.append('url', url);

    if (options.label) {
      params.append('label', options.label);
    }
    if (options.destination) {
      params.append('dir_edit', options.destination);
    }

    return this.callServer({
      path: '/php/addtorrent.php',
      data: params,
    });
  }

  /**
   * Get the list of torrents.
   *
   * @param  {array}          fields
   * @return {Promise<array>}
   */
  get(fields = []) {
    const data = qs.stringify({ mode: 'list' });

    return this.callServer({
      path: '/plugins/httprpc/action.php',
      data,
    }).then((data) => Object.keys(data.t).map((hashString) => {
      const torrent = utils.getTorrentInfo(data.t[hashString]);
      const res = { hashString };
      for (let i = 0 ; i < fields.length ; ++i) {
        res[fields[i]] = torrent[fields[i]];
      }
      return res;
    }));
  }

  /**
   * Delete a torrent from its hash.
   *
   * @param  {string}          hash
   * @param  {boolean}         deleteTiedFiles
   * @return {Promise<object>}
   */
  delete(hash, deleteTiedFiles = true) {
    let data = `<?xml version="1.0" encoding="UTF-8"?>
      <methodCall>
        <methodName>system.multicall</methodName>
        <params>
          <param>
            <value>
              <array>
                <data>
                  <value>
                    <struct>
                      <member>
                        <name>methodName</name>
                        <value>
                          <string>d.set_custom5</string>
                        </value>
                      </member>
                      <member>
                        <name>params</name>
                        <value>
                          <array>
                            <data>
                              <value>
                                <string>${hash}</string>
                              </value>
                              <value>
                                <string>1</string>
                              </value>
                            </data>
                          </array>
                        </value>
                      </member>
                    </struct>
                  </value>`;
    if (deleteTiedFiles) {
      data += `<value>
        <struct>
          <member>
            <name>methodName</name>
            <value>
              <string>d.delete_tied</string>
            </value>
          </member>
          <member>
            <name>params</name>
            <value>
              <array>
                <data>
                  <value>
                    <string>${hash}</string>
                  </value>
                </data>
              </array>
            </value>
          </member>
        </struct>
      </value>`;
    }
    data += `<value>
                    <struct>
                      <member>
                        <name>methodName</name>
                        <value>
                          <string>d.erase</string>
                        </value>
                      </member>
                      <member>
                        <name>params</name>
                        <value>
                          <array>
                            <data>
                              <value>
                                <string>${hash}</string>
                              </value>
                            </data>
                          </array>
                        </value>
                      </member>
                    </struct>
                  </value>
                </data>
              </array>
            </value>
          </param>
        </params>
      </methodCall>`;

    return this.callServer({
      type: 'text/xml; charset=UTF-8',
      path: '/plugins/httprpc/action.php',
      data,
    });
  }
}

module.exports = RuTorrent;


utils.js
module.exports = {
  getTorrentInfo: data => ({
    'd.is_open': data[0],
    'd.is_hash_checking': data[1],
    'd.is_hash_checked': data[2],
    'd.get_state': data[3],
    'd.get_name': data[4],
    'd.get_size_bytes': data[5],
    'd.get_completed_chunks': data[6],
    'd.get_size_chunks': data[7],
    'd.get_bytes_done': data[8],
    'd.get_up_total': data[9],
    'd.get_ratio': data[10],
    'd.get_up_rate': data[11],
    'd.get_down_rate': data[12],
    'd.get_chunk_size': data[13],
    'd.get_custom1': data[14],
    'd.get_peers_accounted': data[15],
    'd.get_peers_not_connected': data[16],
    'd.get_peers_connected': data[17],
    'd.get_peers_complete': data[18],
    'd.get_left_bytes': data[19],
    'd.get_priority': data[20],
    'd.get_state_changed': data[21],
    'd.get_skip_total': data[22],
    'd.get_hashing': data[23],
    'd.get_chunks_hashed': data[24],
    'd.get_base_path': data[25],
    'd.get_creation_date': data[26],
    'd.get_tracker_focus': data[27],
    'd.is_active': data[28],
    'd.get_message': data[29],
    'd.get_custom2': data[30],
    'd.get_free_diskspace': data[31],
    'd.is_private': data[32],
    'd.is_multi_file': data[33],
  }),
};
